File: frontend/pai_mobile/android/app/src/main/kotlin/com/example/pai_mobile/agent/plugins/PluginManager.kt
package com.pai.agent.plugins

class PluginManager(private val registry: NativeModuleRegistry) {

    suspend fun invoke(
        nativeModule: String,
        capability: String,
        params: Map<String, Any?>,
    ): Map<String, Any?> {
        val m = registry.get(nativeModule)
            ?: return mapOf("ok" to false, "error" to "module_not_available")

        if (capability !in m.capabilities) {
            return mapOf("ok" to false, "error" to "capability_not_supported")
        }

        return try {
            m.invoke(capability, params)
        } catch (t: Throwable) {
            mapOf("ok" to false, "error" to (t.message ?: "unknown_error"))
        }
    }

    suspend fun requestPermissions(
        nativeModule: String,
        permissions: List<String>,
    ): Map<String, Any?> {
        val m = registry.get(nativeModule)
            ?: return mapOf("ok" to false, "error" to "module_not_available")
        return try {
            m.requestPermissions(permissions)
        } catch (t: Throwable) {
            mapOf("ok" to false, "error" to (t.message ?: "unknown_error"))
        }
    }
}

File: frontend/pai_mobile/linux/runner/agent/plugins/plugin_manager.cc
// linux/runner/agent/plugins/plugin_manager.cc
#include "plugin_manager.h"

#include <algorithm>

namespace pai::agent {

std::map<std::string, std::string> PluginManager::Invoke(
    const std::string& native_module,
    const std::string& capability,
    const std::map<std::string, std::string>& params) {
  auto m = registry_->Get(native_module);
  if (!m) {
    return {{"ok", "false"}, {"error", "module_not_available"}};
  }

  auto caps = m->Capabilities();
  if (std::find(caps.begin(), caps.end(), capability) == caps.end()) {
    return {{"ok", "false"}, {"error", "capability_not_supported"}};
  }

  try {
    return m->Invoke(capability, params);
  } catch (const std::exception& e) {
    return {{"ok", "false"}, {"error", e.what()}};
  }
}

std::map<std::string, std::string> PluginManager::RequestPermissions(
    const std::string& native_module,
    const std::vector<std::string>& permissions) {
  auto m = registry_->Get(native_module);
  if (!m) {
    return {{"ok", "false"}, {"error", "module_not_available"}};
  }
  try {
    return m->RequestPermissions(permissions);
  } catch (const std::exception& e) {
    return {{"ok", "false"}, {"error", e.what()}};
  }
}

}  // namespace pai::agent